### PR TITLE
MTSRE-499 | add support for syncsetMigration field in addon metadata

### PR DESCRIPTION
**Pull request by yashvardhan-kukreja**
Signed-off-by: Yashvardhan Kukreja <ykukreja@redhat.com>

### CODE REVIEW

**Comment by apahim** [comment]
This looks nice! Still missing the ocm cli implementation part, but the part covered here looks good.

**Comment by yashvardhan-kukreja** [comment]
> Still missing the ocm cli implementation part

ack'd, will take a look at it :)

File: managedtenants/schemas/metadata.schema.yaml [hunk 1]
```diff
@@ -57,6 +57,17 @@ properties:
...
+  syncsetMigration:
+    type: string
+    enum:
+      - "step 1 - block installations"
+      - "step 2 - orphan SSS objects"
+      - "step 3 - change SSS label"
+      - "step 4 - enable syncset"
+      - "step 5 - migration complete"
+      - "rollback ocm"
+      - "rollback selectorsyncset"
```
Comment (sugarraysam) [inline managedtenants/schemas/metadata.schema.yaml:69]: I could not confirm if enum allow space in their character set. Your test pass so it should be ok.

Comment (yashvardhan-kukreja) [reply managedtenants/schemas/metadata.schema.yaml:69]: Thanks for raising this Sam! 
Just confirmed that the enum is being obeyed :)

File: managedtenants/data/selectorsyncset.yaml.j2 [hunk 2]
```diff
@@ -179,8 +187,17 @@ items:
   spec:
     clusterDeploymentSelector:
       matchLabels:
+        {% if ADDON.metadata.get('syncsetMigration') in ["step 3 - change SSS label", "step 4 - enable syncset", "step 5 - migration complete", "rollback ocm"] %}
```
Comment (ashishmax31) [inline managedtenants/data/selectorsyncset.yaml.j2:190]: As discussed, we don't need to change this section.

**Comment by ashishmax31** [comment]
This looks good, please remove the change under the `ADDON.manager == AddonManager.ADDON_OPERATOR` section.

File: managedtenants/schemas/metadata.schema.yaml (hunk 1 above)
Comment (yashvardhan-kukreja) [reply managedtenants/schemas/metadata.schema.yaml:69]: Added a test to validate false positives as well

File: managedtenants/data/selectorsyncset.yaml.j2 (hunk 2 above)
Comment (yashvardhan-kukreja) [reply managedtenants/data/selectorsyncset.yaml.j2]: addressed :)

**Comment by yashvardhan-kukreja** [comment]
Verified the OCM API implementation hands-on and made the OcmCli package consistent with it.

File: managedtenants/utils/ocm.py [hunk 3]
```diff
@@ -23,7 +23,7 @@ def retry_hook(exception):
         raise exception
 
 
-class OcmCli:
+class OcmCli:  # pylint: disable=R0904
```
Comment (apahim) [inline managedtenants/utils/ocm.py:26]: I believe it's safe to disable the `too-many-public-methods` globally https://github.com/mt-sre/managed-tenants-cli/blob/main/.linters/pylint#L83

File: managedtenants/utils/ocm.py [hunk 4]
```diff
@@ -159,6 +159,83 @@ def disable_addon(self, addon_id):
             f"/api/clusters_mgmt/v1/addons/{addon_id}", json={"enabled": False}
         )
 
+    def get_migrations(self):
+        return self._get("/api/clusters_mgmt/v1/addon_migrations")
+
+    def get_migration(self, addon_id):
+        return self._get(f"/api/clusters_mgmt/v1/addon_migrations/${addon_id}")
+
+    def post_addon_migration(self, addon_id):
```
Comment (apahim) [inline managedtenants/utils/ocm.py:168]: get_migration vs post_addon_migration... let's use the same naming. either:

- get_migration and post_migration
or
- get_addon_migration and post_addon_migration

File: managedtenants/utils/ocm.py [hunk 5]
```diff
@@ -159,6 +159,83 @@ def disable_addon(self, addon_id):
...
+        return self.patch_addon_migration(
+            addon_id,
+            {
+                "enabled": True,
+                "white_listed": False,
+                "rollback_migration": False,
+            },
+        )
+
+    def mark_addon_migration_complete(self, addon_id):
```
Comment (apahim) [inline managedtenants/utils/ocm.py:212]: suggestion: `complete_addon_migration`

File: managedtenants/utils/ocm.py [hunk 6]
```diff
@@ -159,6 +159,83 @@ def disable_addon(self, addon_id):
...
+        return self.patch_addon_migration(
+            addon_id,
+            {
+                "enabled": True,
+                "white_listed": False,
+                "rollback_migration": True,
+            },
+        )
+
+    def unrollback_addon_migration(self, addon_id):
```
Comment (apahim) [inline managedtenants/utils/ocm.py:231]: Isn't this the same as `post_addon_migration`?

**Review by apahim — APPROVED** [review]
lgtm. A few nits re naming, but overall looks nice

File: managedtenants/utils/ocm.py (hunk 6 above)
Comment (yashvardhan-kukreja) [reply managedtenants/utils/ocm.py:231]: The `post_addon_migration` one creates the addon_migration object whereas this one patches it :)

File: managedtenants/utils/ocm.py (hunk 3 above)
Comment (yashvardhan-kukreja) [reply managedtenants/utils/ocm.py:26]: Sure, on it :)

**Pull request merged**